### DISCUSSION TITLE
A conceptual question about applying derived data to validation

### DISCUSSION THREAD

**Discussion by gilbox** [day8/re-frame#705] [Q&A]
Hi, this question is not about re-frame but rather about derived data in general and how you would apply it to validation.

Say I have some initial app state that looks like this:

```
user: {
  name: ____,
  phone: ____,
  age: ____,
}
```

For validation each field has the concept of being **dirty** and **valid**. Based on these two states, the field has various representations.
- **not dirty** and **not valid** --> no outline
- **not dirty** and **valid** --> no outline
- **dirty** and **not valid** --> red outline
- **dirty** and **valid** --> blue outline

Representing **valid** using derived data should be trivial. However, how do you deal with **dirty**? The way I have been doing it is by annotating my data like this:

```
user: {
  name: { value: 'gilbox', dirty: true },
  phone: { value: '', dirty: false },
  age: { value: 99, dirty: true },
}
```

IOW, **dirty** is not derived data, but **valid** is derived.

However, is there another way of modeling this so that the data looks like the following and the concept of **dirty** is either derived or stored elsewhere? Is this even a good idea?

```
user: {
  name: 'gilbox',
  phone: '',
  age: 99,
}
```

btw, the code I'm working with is [here](https://github.com/gilbox/elegant-react/blob/validation-derived-data/examples/form-validation/UserForm.js#L11) and [here](https://github.com/gilbox/elegant-react/blob/validation-derived-data/examples/form-validation/ValidatingInput.js#L25)

**Comment by ul** [comment]
I usually have a set of dirty field paths under separate key in app-db. Another solution is to keep snapshot of synced values and derive dirty by compare that values with input's current ones.

**Comment by mike-thompson-day8** [comment]
Middleware is perfect for this sort of stuff.  

Remember a middleware can see the "before" and "after" values, to decide if something has changed - become dirty.  Ie. before the event handler ran, and after.  So it can tell if the event handler changed a value.  And then set a dirty flag somewhere (I don't have a strong opinion about where).   

Then, another piece of middleware (or the same one as above) can set valid flags.  This is particularly powerful when the valid flag is a function of multiple values, although I think here you just want to validate each value individually.  Anyway, the validation code need only be run when a value changes (see above). 

For an example of what's possible, look at the `on-change` middleware
https://github.com/Day8/re-frame/blob/78dacf5ca1551b25b53cc527fbe4b26d3de1ceaf/src/re_frame/middleware.cljs#L198-L236:

**Comment by gilbox** [comment]
Thanks for the ideas guys, fascinating stuff.

Mike, is there any reason you'd prefer middleware over a reaction for validation?

btw, I enjoyed reviewing your documentation on middleware, it's very well-written.

**Comment by mike-thompson-day8** [comment]
You are right to say you have a choice around validation:  
1. a subscription handler can calculate validation state, and give it to the view
2. event handler middleware can calculate validation state - and then assoc it into `app-db` somewhere. 

On the one hand, I have a mild preference for path 2 because it strikes me as a bit easier to debug and  will be more efficient if the validation process is complicated. 

On the other hand, path 1 can never get out of step.  So there's clearly an argument that way too.  

Back to the original hand, path 2 is useful if things other than views need to know about the current validation state (other event handlers?). 

So I'm pretty ambidextrous on this issue.  I think it depends on context.

**Comment by mike-thompson-day8** [comment]
@gilbox  are you happy to close this?  Do you have a satisfactory solution for the `:dirty?` and `:valid?` computations?

**Comment by gilbox** [comment]
Yes, thanks :)
